flyteplugins/go/tasks/pluginmachinery/flytek8s/pod_helper.go

```diff
@@ -48,31 +96,131 @@ func ApplyInterruptibleNodeSelectorRequirement(interruptible bool, affinity *v1.
+		new = proto.Clone(base).(*core.ExtendedResources)
+	}
+
+	// No overrides found
+	if overrides == nil {
+		return new
+	}
+
+	// GPU Accelerator
+	if overrides.GetGpuAccelerator() != nil {
+		new.GpuAccelerator = overrides.GetGpuAccelerator()
 	}
 
+	return new
 }
 
-// ApplyInterruptibleNodeAffinity configures the node-affinity for the pod using the configuration specified.
-func ApplyInterruptibleNodeAffinity(interruptible bool, podSpec *v1.PodSpec) {
+func ApplyGPUNodeSelectors(podSpec *v1.PodSpec, gpuAccelerator *core.GPUAccelerator) {
+	// Short circuit if pod spec does not contain any containers that use GPUs
+	requiresGPUs := false
+	for _, cnt := range podSpec.Containers {
+		if _, ok := cnt.Resources.Limits[config.GetK8sPluginConfig().GpuResourceName]; ok {
+			requiresGPUs = true
+			break
+		}
+	}
+	if !requiresGPUs {
+		return
+	}
+
 	if podSpec.Affinity == nil {
 		podSpec.Affinity = &v1.Affinity{}
 	}
 
-	ApplyInterruptibleNodeSelectorRequirement(interruptible, podSpec.Affinity)
+	// Apply changes for GPU device
+	device := gpuAccelerator.GetDevice()
+	if len(device) > 0 {
+		// Add node selector requirement for GPU device
+		deviceNsr := v1.NodeSelectorRequirement{
+			Key:      config.GetK8sPluginConfig().GpuDeviceNodeLabel,
+			Operator: v1.NodeSelectorOpIn,
+			Values:   []string{device},
+		}
+		AddRequiredNodeSelectorRequirements(podSpec.Affinity, deviceNsr)
+		// Add toleration for GPU device
+		deviceTol := v1.Toleration{
+			Key:      config.GetK8sPluginConfig().GpuDeviceNodeLabel,
+			Value:    device,
+			Operator: v1.TolerationOpEqual,
+			Effect:   v1.TaintEffectNoSchedule,
+		}
+		podSpec.Tolerations = append(podSpec.Tolerations, deviceTol)
+	}
+
+	// Short circuit if a partition size preference is not specified
+	partitionSizeValue := gpuAccelerator.GetPartitionSizeValue()
+	if partitionSizeValue == nil {
+		return
+	}
+
+	// Apply changes for GPU partition size, if applicable
+	var partitionSizeNsr *v1.NodeSelectorRequirement
+	var partitionSizeTol *v1.Toleration
+	switch partitionSizeValue.(type) {
+	case *core.GPUAccelerator_Unpartitioned:
+		if !gpuAccelerator.GetUnpartitioned() {
+			break
+		}
+		if config.GetK8sPluginConfig().GpuUnpartitionedNodeSelectorRequirement != nil {
+			partitionSizeNsr = config.GetK8sPluginConfig().GpuUnpartitionedNodeSelectorRequirement
+		} else {
+			partitionSizeNsr = &v1.NodeSelectorRequirement{
+				Key:      config.GetK8sPluginConfig().GpuPartitionSizeNodeLabel,
+				Operator: v1.NodeSelectorOpDoesNotExist,
+			}
+		}
+		if config.GetK8sPluginConfig().GpuUnpartitionedToleration != nil {
+			partitionSizeTol = config.GetK8sPluginConfig().GpuUnpartitionedToleration
+		} else {
+			partitionSizeTol = &v1.Toleration{
+				Key:      config.GetK8sPluginConfig().GpuPartitionSizeNodeLabel,
+				Value:    GpuPartitionSizeNotSet,
```

Comment:
Is that a k8s const value to mean something?

Reply:
No, it’s an arbitrary default value that we set for the toleration value added to tasks that explicitly require unpartitioned GPUs. See: https://github.com/flyteorg/flyte/blob/ee11093a396d77f818b1bfe5752970e3388e4d4a/flyteplugins/go/tasks/pluginmachinery/flytek8s/pod_helper.go#L36

The toleration is useful if running multiple GPU node pools and an unpartitioned A100 node group should be protected. The toleration can also be overridden via plugin config.

Reply:
Do you think this can be a surprising behavior?

Reply:
Possibly. We can either:
1. Document this behavior. An extraneous toleration should generally be a no-op, unless it somehow conflicts with a real, but unrelated, taint.
2. Drop the "default" toleration and require that it be specified explicitly by config, if unpartitioned GPU node groups are protected.

Reply:
@EngHabu: Does (2) seem reasonable to you?

Reply:
Went ahead and made the change! :)